File: 2. Structural Design Patterns/Facade/Python/FacadeExample.py
# Facade Pattern Example in Python
# Subsystems: CPU, Memory, HardDrive; Facade: Computer

class CPU:
    def freeze(self):
        return "CPU: freeze"
    
    def jump(self, position):
        return f"CPU: jump to {position}"
    
    def execute(self):
        return "CPU: execute"

class Memory:
    def load(self, position, data):
        return f"Memory: load {len(data)} bytes at {position}"

class HardDrive:
    def read(self, lba, size):
        return f"HardDrive: read {size} bytes from LBA {lba}"

class Computer:
    def __init__(self):
        self.cpu = CPU()
        self.memory = Memory()
        self.hard_drive = HardDrive()
    
    def start(self):
        results = []
        results.append(self.cpu.freeze())
        boot_data = [0] * 1024  # dummy boot data
        results.append(self.hard_drive.read(0, 1024))
        results.append(self.memory.load(0, boot_data))
        results.append(self.cpu.jump(0))
        results.append(self.cpu.execute())
        return "\n".join(results)
    
    def shutdown(self):
        return "Computer: shutdown sequence"

def main():
    computer = Computer()
    print(computer.start())
    print(computer.shutdown())

if __name__ == "__main__":
    main()

File: 2. Structural Design Patterns/Facade/Kotlin/FacadeExample.kt
// Facade Pattern Example in Kotlin
// Subsystems: CPU, Memory, HardDrive; Facade: Computer

class CPU { fun freeze()=println("CPU: freeze"); fun jump(position: Long)=println("CPU: jump to $position"); fun execute()=println("CPU: execute") }
class Memory { fun load(position: Long, data: ByteArray)=println("Memory: load ${data.size} bytes at $position") }
class HardDrive { fun read(lba: Long, size: Int): ByteArray { println("HardDrive: read $size bytes from LBA $lba"); return ByteArray(size) } }

class Computer {
    private val cpu = CPU(); private val memory = Memory(); private val hardDrive = HardDrive()
    fun start(){ cpu.freeze(); val boot = hardDrive.read(0,1024); memory.load(0, boot); cpu.jump(0); cpu.execute() }
    fun shutdown(){ println("Computer: shutdown sequence") }
}

fun main(){ val c = Computer(); c.start(); c.shutdown() }
